tools/cell_census_builder/validate.py

```diff
@@ -66,101 +66,115 @@ def validate_all_soma_objects_exist(soma_path: str, experiment_builders: List[Ex
-        assert "X" in rna and rna["X"].exists() and rna["X"].soma_type == "SOMACollection"
-        for lyr in CENSUS_X_LAYERS:
-            # layers only exist if there are cells in the measurement
-            if lyr in rna.X:
-                assert rna.X[lyr].exists() and rna.X[lyr].soma_type == "SOMASparseNDArray"
-
-        # and a dataset presence matrix
-        # dataset presence only exists if there are cells in the measurement
-        if FEATURE_DATASET_PRESENCE_MATRIX_NAME in rna:
-            assert rna[FEATURE_DATASET_PRESENCE_MATRIX_NAME].exists()
-            assert rna[FEATURE_DATASET_PRESENCE_MATRIX_NAME].soma_type == "SOMASparseNDArray"
-            # TODO(atolopko): validate 1) shape, 2) joinids exist in datsets and var
+        # there should be an experiment for each builder
+        census_data = census[CENSUS_DATA_NAME]
+        for eb in experiment_builders:
+            assert eb.name in census_data
+            assert soma.Experiment.exists(census_data[eb.name].uri)
+            assert census_data[eb.name].soma_type == "SOMAExperiment"
+
+            e = census_data[eb.name]
+            assert "obs" in e
+            assert soma.DataFrame.exists(e.obs.uri)
+            assert e.obs.soma_type == "SOMADataFrame"
+            assert "ms" in e
+            assert soma.Collection.exists(e.ms.uri)
+            assert e.ms.soma_type == "SOMACollection"
+
+            # there should be a single measurement called 'RNA'
+            assert "RNA" in e.ms
+            assert soma.Measurement.exists(e.ms["RNA"].uri)
+            assert e.ms["RNA"].soma_type == "SOMAMeasurement"
+
+            # The measurement should contain all X layers where n_obs > 0 (existence checked elsewhere)
+            rna = e.ms["RNA"]
+            assert "var" in rna
+            assert soma.DataFrame.exists(rna["var"].uri)
+            assert rna["var"].soma_type == "SOMADataFrame"
+            assert "X" in rna
+            assert soma.Collection.exists(rna["X"].uri)
+            assert rna["X"].soma_type == "SOMACollection"
+            for lyr in CENSUS_X_LAYERS:
+                # layers only exist if there are cells in the measurement
+                if lyr in rna.X:
+                    assert soma.SparseNDArray.exists(rna.X[lyr].uri)
+                    assert rna.X[lyr].soma_type == "SOMASparseNDArray"
+
+            # and a dataset presence matrix
+            # dataset presence only exists if there are cells in the measurement
+            if eb.n_obs > 0:
+                assert FEATURE_DATASET_PRESENCE_MATRIX_NAME in rna
+                assert soma.SparseNDArray.exists(rna[FEATURE_DATASET_PRESENCE_MATRIX_NAME].uri)
+                assert rna[FEATURE_DATASET_PRESENCE_MATRIX_NAME].soma_type == "SOMASparseNDArray"
+                presence = rna[FEATURE_DATASET_PRESENCE_MATRIX_NAME].read().tables().concat().to_pandas()
+                assert len(presence) > 0
```

Comment:
I did add this one line to assert presence has some data (caught a bug in refactoring)

Reply:
any reason to not add the same check for obs/var/X?

You can also use `nnz` to make it faster, rather than having to read the data as you did here.

Reply:
`nnz` now in use! (didn't know about it)

I think I started to add checks for presence shape, and it started to take time, so I punted and left the below TODO

Reply:
I do see `nnz` available anywhere, but we can do this:
```
                assert sum([c.non_zero_length for c in rna['feature_dataset_presence_matrix'].read().csrs()]) > 0
```
in case this saves time instantiating as a pandas df.